src/core/mesh/qgsmeshtracerenderer.cpp

```diff
@@ -0,0 +1,605 @@
+
+  if ( ! mActiveFaceFlagValue.active( mTriangularMesh.trianglesToNativeFaces()[mCacheFaceIndex] ) )
+    return ( QgsVector( std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN() ) );
+
+  face = mTriangularMesh.triangles().at( mCacheFaceIndex );
+
+
+  QgsPoint p1 = mTriangularMesh.vertices().at( face[0] );
+  QgsPoint p2 = mTriangularMesh.vertices().at( face[1] );
+  QgsPoint p3 = mTriangularMesh.vertices().at( face[2] );
+
+  QgsVector v1 = QgsVector( mDatasetValues.value( face[0] ).x(),
+                            mDatasetValues.value( face[0] ).y() );
+
+  QgsVector v2 = QgsVector( mDatasetValues.value( face[1] ).x(),
+                            mDatasetValues.value( face[1] ).y() );
+
+  QgsVector v3 = QgsVector( mDatasetValues.value( face[2] ).x(),
+                            mDatasetValues.value( face[2] ).y() );
+
+
+  return QgsMeshLayerUtils::interpolateVectorFromVerticesData(
+           p1,
+           p2,
+           p3,
+           v1,
+           v2,
+           v3,
+           point );
+}
+
+QgsMeshVectorValueInterpolator::QgsMeshVectorValueInterpolator( QgsTriangularMesh triangularMesh,
+    QgsMeshDataBlock datasetVectorValues,
+    QgsMeshDataBlock scalarActiveFaceFlagValues ):
+  mTriangularMesh( triangularMesh ),
+  mDatasetValues( datasetVectorValues ),
+  mActiveFaceFlagValue( scalarActiveFaceFlagValues )
+{
+
+}
+
+QgsMeshVectorValueInterpolator::~QgsMeshVectorValueInterpolator() {}
+
+void QgsMeshVectorValueInterpolator::updateFace( const QgsPointXY &point )
+{
+  if ( ! QgsMeshUtils::isInTriangleFace( point, face, mTriangularMesh.vertices() ) )
+  {
+    /*the first version of this method uses QgsGeometry::contain to check if the point is in triangles,
+    *  not too time consuming for a simple triangle?
+    * why not using simpler function for this check ?  version 2 is  nearly 10 time faster.
+    */
+    mCacheFaceIndex = mTriangularMesh.faceIndexForPoint_v2( point );
+  }
+
+}
+
+QSize QgsMeshTraceField::size() const
+{
+  return mFieldSize;
+}
+
+QPoint QgsMeshTraceField::topLeft() const
+{
+  return mFieldTopLeftInDeviceCoordinates;
+}
+
+int QgsMeshTraceField::particleWidth() const
+{
+  return mFieldResolution;
+}
+
+QgsPointXY QgsMeshTraceField::positionToMapCoordinnate( const QPoint &pixelPosition, const QgsPointXY &positionInPixel )
+{
+  QgsPointXY mapPoint = mMapToFieldPixel.toMapCoordinates( pixelPosition );
+  mapPoint = mapPoint + QgsVector( positionInPixel.x() * mMapToFieldPixel.mapUnitsPerPixel(),
+                                   positionInPixel.y() * mMapToFieldPixel.mapUnitsPerPixel() );
+  return mapPoint;
+}
+
+QgsMeshTraceField::QgsMeshTraceField( const QgsRenderContext &renderContext, QgsMeshVectorValueInterpolator *interpolator, const QgsRectangle &layerExtent, double Vmax ):
+  mLayerExtent( layerExtent ),
+  mRenderContext( renderContext ),
+  mVectorValueInterpolator( interpolator ),
+  mVmax( Vmax )
+{
+}
+
+void QgsMeshTraceField::updateSize()
```

Comment:
can you please describe what this function does? I do not exactly know what are you trying to achieve?

if you need pixel coords for face, look how it is done in https://github.com/qgis/QGIS/blob/461d2edac3f092b636f114d9dbf8ec4b84e89aa8/src/core/mesh/qgsmeshlayerinterpolator.cpp#L108 or https://github.com/qgis/QGIS/blob/461d2edac3f092b636f114d9dbf8ec4b84e89aa8/src/core/mesh/qgsmeshvectorrenderer.cpp#L48

Reply:
To calculate stream line and render it, there is a pixels fields and a resolution of this pixel field. 
The resolution is the number of device pixels per pixel field. Increasing the resolution permits to reduce the calculation time. 
This functions calculates the size of the pixels field depending of the resolution, but also depending of the mesh extent. If the mesh doesn't occupy all the extent device, no need to have a pixel field outside the mesh extent. This approach reduces the calculation time.